internal/contour/handler.go

```diff
@@ -164,11 +181,28 @@ func (e *EventHandler) Start(ctx context.Context) error {
 				// not to process it.
 				e.incSequence()
 			}
+
+			// We're done processing this event
+			if updateOpAdd, ok := op.(opAdd); ok {
+				if updateOpAdd.isInInitialList {
+					e.syncTracker.Finished()
+				}
+			}
 		case <-pending:
+			// Ensure informer caches are synced.
+			// Schedule a retry for dag rebuild if cache is not synced yet.
+			// Note that we can't block and wait for the cache sync as it depends on progress of this loop.
+			if !e.syncTracker.HasSynced() {
+				e.Info("skipping dag rebuild as cache is not synced")
+				timer.Reset(e.holdoffDelay)
+				break
+			}
+
 			e.WithField("last_update", time.Since(lastDAGRebuild)).WithField("outstanding", reset()).Info("performing delayed update")
 			e.rebuildDAG()
 			e.incSequence()
 			lastDAGRebuild = time.Now()
+			e.initialDagBuilt = true
```

Comment:
Maybe something like this

```diff
diff --git a/internal/contour/handler.go b/internal/contour/handler.go
index 76db0a49..716bf982 100644
--- a/internal/contour/handler.go
+++ b/internal/contour/handler.go
@@ -199,10 +199,21 @@ func (e *EventHandler) Start(ctx context.Context) error {
                        }

                        e.WithField("last_update", time.Since(lastDAGRebuild)).WithField("outstanding", reset()).Info("performing delayed update")
-                       e.rebuildDAG()
+
+                       // Build a new DAG and sends it to the Observer.
+                       latestDAG := e.builder.Build()
+                       e.observer.OnChange(latestDAG)
+
+                       // Allow XDS server to start (if it hasn't already).
+                       e.initialDagBuilt = true
+
+                       // Update the status on objects.
+                       for _, upd := range latestDAG.StatusCache.GetStatusUpdates() {
+                               e.statusUpdater.Send(upd)
+                       }
+
                        e.incSequence()
                        lastDAGRebuild = time.Now()
-                       e.initialDagBuilt = true
                case <-ctx.Done():
                        // shutdown
                        return nil
@@ -271,14 +282,3 @@ func (e *EventHandler) incSequence() {
        default:
        }
 }
-
-// rebuildDAG builds a new DAG and sends it to the Observer,
-// the updates the status on objects, and updates the metrics.
-func (e *EventHandler) rebuildDAG() {
-       latestDAG := e.builder.Build()
-       e.observer.OnChange(latestDAG)
-
-       for _, upd := range latestDAG.StatusCache.GetStatusUpdates() {
-               e.statusUpdater.Send(upd)
-       }
-}
```

Here I inlined the function since I felt like putting `e.initialDagBuilt = true`  inside `rebuildDAG` is somehow hiding the fact, but this was just my thinking.

This will allow XDS to be served, so Envoys will be serving traffic long before statuses are up to date.

Reply:
Oh! Sorry for the misinterpretation 😅. You're absolutely right.

Reply:
Pushed a new commit to fix this.